url47ebb27197_asirbu_CorrectiveEvolution_tgz-pJ8-eu_fbk_soa_evolution_engine_impl_test_ProblemToSMVFuncTestJ8/src/eu/fbk/soa/evolution/engine/impl/STSToSMV.java

```diff
@@ -0,0 +1,356 @@
+						remainingTrans.remove(trans2);
+					}
+				}
+				transitionGroups.add(newGroup);
+			}
+		}
+			
+		for (Set<Transition> trGroup : transitionGroups) {
+			translation += translateTransitionGroup(trGroup, translatedEntity);
+		}
+		return translation;
+	}
+
+
+	private String getActionDeclaration(Set<Action> totalActions, 
+			Object translatedEntity, String type) {
+		String translation = "";
+		
+		Set<Action> actions = new HashSet<Action>(totalActions);
+		for (Action act : totalActions) {
+			if (!act.getRelatedEntity().equals(translatedEntity)) {
+				actions.remove(act);
+			}
+		}
+		if (!actions.isEmpty()) { 
+			translation += "IVAR " + type + ": {UNDEF";
+			Set<Action> noUndefActions = new HashSet<Action>(totalActions);
+			for (Action act : totalActions) {
+				if (act.getName().equals("UNDEF")) {
+					noUndefActions.remove(act);
+				}
+			}
+			String actionsDecl = StringUtils.getCommaSeparatedString(noUndefActions);
+			if (actionsDecl != "") {
+				translation += ", " + actionsDecl;
+			}
+			translation += "};\n";
+		}
+		return translation;
+	}
+	
+	private String translateTransitionGroup(Set<Transition> trGroup,
+			Object translatedEntity) {
+		
+		String translation = "";
+		if (trGroup.size() >= 1) {
+			Transition trans = trGroup.iterator().next();
+			
+			translation +=
+				this.translateTransitionCondition(trans, translatedEntity);
+			translation += "& " + translateAction(trans.getAction(), translatedEntity) + " ";	
+			
+			translation += ": ";
+			
+			int nr = 0;
+			String targetStates = "";
+			for (Transition t : trGroup) {
+				State target = t.getTarget();
+				if (nr > 0) {
+					targetStates += ", ";
+				}
+				targetStates += target.getName();
+				nr++;
+			}
+			if (nr == 1) {
+				translation += targetStates;
+			} else {
+				translation += "{" + targetStates + "}";
+			}
+			
+			translation += ";\n";
+		}
+		return translation;
+	}
+	
+	private String translateTransition(Transition trans, Object translatedEntity) {
+		String translation = 
+			this.translateTransitionCondition(trans, translatedEntity);
+		
+		translation += "& " + translateAction(trans.getAction(), translatedEntity) + " ";	
+		
+		translation += ": " + trans.getTarget() + ";\n";
+		return translation;	
+	}
+	
+	private String translateTransitionCondition(Transition trans, Object translatedEntity) {
+		String translation = "(state = " + trans.getSource() + ") ";
+		
+		String formula = translateStateFormula(trans.getCondition());
+		if (formula != "") {
+			translation += "& " + formula;
+		}
+		
+		for (Clause<? extends Literal> clause : trans.getGuardClauses()) {
+			String clauseCond = "";
+			Iterator<? extends Literal> iterator = clause.getLiterals().iterator();
+			
+			while (iterator.hasNext()) {
+				Literal lit = iterator.next();
+				String guardCond = "";
+				if (lit.isNegated()) {
+					guardCond += "!(";
+				}
+				for (String entityID : id2STS.keySet()) {
+					STS sts = id2STS.get(entityID);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: Iterating over [map.keySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#keySet--) and calling [map.get(Key)](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#get-java.lang.Object-) to get the values within the loop is inefficient. 
**Fix**: Iterate over [map.entrySet()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#entrySet--) and access the values using [entry.getValue()](https://docs.oracle.com/javase/8/docs/api/java/util/Map.Entry.html#getValue--) to avoid potential performance penalty due to lookup.